
August 2015  
Modified September 2017, October 2018

Author: FBB

** reproduce the result in **
http://www.mdrc.org/sites/default/files/What%20Strategies%20Work%20for%20the%20Hard%20FR.pdf 

In [1]:
from __future__ import print_function, division
__author__ = 'uak211'
import numpy as np

### data on those who were ever employed in a CEO job

null hypothesis: the % of former prisoners employed after release is the same or lower for candidates who participated in the program as for the control group, significance level p = 0.05

$H_0: P_0 - P_1 \geq$ 0 
    
$H_a: P_0 - P_1 $< 0    
    
    
 $\alpha$ = 0.05    

this is a TEST OF PROPORTIONS. we use the Binomial distribution since it is a yes/no (bernulli) test for each subject: the former inmate was or was not ever employed in a CEO transitional job: 

** $P_0=0.035, P_1=0.701$**

#### z-score

In [2]:
### from Dr. Bianco's homework notebook ###

# first thing: choose your significance threshold
alpha = 0.05
# we like fractions better then percentages. as a rule of thumb, either use fractions or counts. 
# 0 for control group, 1 for program group

P_0 = 3.5 * 0.01 
P_1 = 70.1 * 0.01

if P_0 - P_1 >= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

#sample sizes 

n_0 = 409
n_1 = 564
    
#lets get the counts by multiplying by the sample size

Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1

p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
# standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1)) 

zscore = lambda p0, p1, s : (p0 - p1) / s
z_2y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print ('z-score: ', z_2y)

we must assess the statistical significance
z-score:  20.7697865408


In [3]:
# function copied from Dr. Bianco's homework repo 
# p-value for testing null hypothesis
p_2y = 1 - 0.9984


def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )

    
report_result(p_2y, alpha)

is the p value 0.00 smaller than the critical value 0.05?
YES!
the Null hypothesis is rejected


#### $\chi^2$ test of proportions

In [4]:
# function copied from Dr. Bianco's homework repo
def evalChisq(values):
    '''Evaluates the chi sq from a contingency value
    Arguments:
    values: 2x2 array or list, the contingengy table
    '''
    if not (len(values.shape) == 2 and values.shape == (2,2)):
        print ("must pass a 2D array")
        return -1
    values = np.array(values)
    E = np.empty_like(values)
    for j in range(len(values[0])):
        for i in range(2):
            
            E[i][j] = ((values[i,:].sum() * values[:,j].sum()) / 
                        (values).sum())
    return ((values - E)**2 / E).sum()

In [5]:
Ntot = 973 # a + b + c + d = tot

sample_values = np.array([[0.701 * 564, 0.299 * 564], [0.0305 * 409, 0.965 * 409]])

print (evalChisq(sample_values))

436.223462575



|employed in a subsidized job |  emplopyed   |not employed |        total             
|-----------------------------|--------------|-------------|----------------|
| test sample                 |0.701 * 564   | 0.299 * 564 |      564       |
| control sample              |0.0305 * 409  | 0.965 * 409 |      409       |
|                             |              |             |                |
| total                       |  407.839     |  563.321    |      973       |

Degrees of freedom: (nrow - 1)(ncol - 1)  = 1 * 1  = 1

Using the standard $\chi^2$ distribution table with degrees of freedom = 1 and $\alpha$ = 0.05, the $\chi^2$ critical value is 3.84. Because the calculated $\chi^2$ value is 436.22, which is more than 3.84, the difference in the two proportions is  significant and we **can** reject the null hypothesis.

### data on those convicted of a felony

null hypothesis:
the percentage of former prisoners convicted of a felony employed after release is the same or lower for candidates who participated in the program as for the control group, with $\alpha$ = 0.05
 
$H_0$: $P_0 - P_1 >= 0$ 
 
$H_a$: $P_0 - P_1 < 0$

significance level: $\alpha$ = 0.05



$P_0$ = 0.117 (control group)

$P_1$ = 0.10 (program group)


#### z-score

In [6]:
p0 = 11.7 * 0.01 # percentage of control group convicted of felony
p1 = 10.0 * 0.01 # percentage of program group convicted of felony

if p0 - p1 >= 0:
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

# calculating z-score anyways
n0 = 409
n1 = 568

p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
# standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1))
zscore = lambda p0, p1, s : (p0 - p1) / s

# z-score 
z_3y = zscore(p1, p0, se(p(p0, p1, n0, n1), n0, n1))
print ('z-score: ', abs(z_3y))

the Null holds
z-score:  0.84764020522


In [7]:
p_3y = 1 - 0.7995 # for value at table row = 0.8, table col = 0.04

report_result(p_3y, alpha)

is the p value 0.20 smaller than the critical value 0.05?
NO!
the Null hypothesis is not rejected


#### $\chi^2$ test of proportions

In [8]:
n_total = 568 + 409
sample_values = np.array([[0.100 * 568, 0.90 * 568], [0.117 * 409, 0.883 * 409]])
print (evalChisq(sample_values))

0.718493917505


#### contingency table


|convicted of a felony      |     yes      |    no       |        total             
|---------------------------|--------------|-------------|----------------|
| test sample               |0.1 * 568     | 0.9 * 568   |      568       |
| control sample            |0.117 * 409   | 0.883 * 409 |      409       |
|                           |              |             |                |
| total                     | 104.653      | 872.347     |       977      |

Degrees of freedom: (nrow - 1)(ncol - 1)  = 1 * 1  = 1

Using the standard $\chi^2$ distribution table with degrees of freedom = 1 and $\alpha$ = 0.05, the $\chi^2$ critical value is 3.84. Because the calculated $\chi^2$ value is 0.718, which is less than 3.84, the difference in the two proportions is not significant and we **cannot** reject the null hypothesis.